File: 3 course/6 semester/Software development for mobile platforms/films_viewer_kotlin/app/src/main/java/com/example/films_viewer_kotlin/services/MovieService.kt
package com.example.films_viewer_kotlin.services

import com.example.films_viewer_kotlin.models.Movie
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.tasks.await

class MovieService {

    private val firestore = FirebaseFirestore.getInstance()

    fun getMoviesStream(): Flow<List<Movie>> = callbackFlow {
        val collectionRef = firestore.collection("movies")
        val listenerRegistration = collectionRef.addSnapshotListener { snapshot, error ->
            if (error != null) {
                close(error)
                return@addSnapshotListener
            }
            val movies = snapshot?.documents?.map { doc ->
                Movie.fromFirestore(doc.data ?: emptyMap(), doc.id)
            } ?: emptyList()
            trySend(movies)
        }
        awaitClose { listenerRegistration.remove() }
    }

    fun getFavoriteMovies(userId: String): Flow<List<Movie>> = callbackFlow {
        val query = firestore.collection("movies")
            .whereArrayContains("favoritedBy", userId)
        val listenerRegistration = query.addSnapshotListener { snapshot, error ->
            if (error != null) {
                close(error)
                return@addSnapshotListener
            }
            val movies = snapshot?.documents?.map { doc ->
                Movie.fromFirestore(doc.data ?: emptyMap(), doc.id)
            } ?: emptyList()
            trySend(movies)
        }
        awaitClose { listenerRegistration.remove() }
    }

    suspend fun toggleFavorite(movieId: String, userId: String) {
        val docRef = firestore.collection("movies").document(movieId)
        val snapshot = docRef.get().await()
        if (!snapshot.exists()) return

        val favoritedBy = snapshot.get("favoritedBy") as? List<*> ?: emptyList<Any>()
        val isFavorite = favoritedBy.contains(userId)

        if (isFavorite) {
            docRef.update("favoritedBy", FieldValue.arrayRemove(userId)).await()
        } else {
            docRef.update("favoritedBy", FieldValue.arrayUnion(userId)).await()
        }
    }

}

File: 3 course/6 semester/Software development for mobile platforms/films_viewer/lib/Services/MovieService.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../Models/Movie.dart';

class MovieService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  // Получение списка всех фильмов
  Stream<List<Movie>> getMoviesStream() {
    return _firestore.collection('movies').snapshots().map((snapshot) {
      return snapshot.docs.map((doc) {
        return Movie.fromFirestore(doc.data(), doc.id);
      }).toList();
    });
  }

  // Получение избранных фильмов для конкретного пользователя
  Stream<List<Movie>> getFavoriteMovies(String userId) {
    return _firestore
        .collection('movies')
        .where('favoritedBy', arrayContains: userId) // Фильтруем по пользователю
        .snapshots()
        .map((snapshot) {
      return snapshot.docs.map((doc) {
        return Movie.fromFirestore(doc.data(), doc.id);
      }).toList();
    });
  }

  // Добавление фильма в избранное
  Future<void> toggleFavorite(String movieId, String userId) async {
    final docRef = _firestore.collection('movies').doc(movieId);
    final doc = await docRef.get();

    if (!doc.exists) return;

    final isFavorite = (doc['favoritedBy'] as List).contains(userId);

    if (isFavorite) {
      await docRef.update({
        'favoritedBy': FieldValue.arrayRemove([userId])
      });
    } else {
      await docRef.update({
        'favoritedBy': FieldValue.arrayUnion([userId])
      });
    }
  }

  // Добавление нового фильма (для админ-панели)
  Future<void> addMovie(Movie movie) async {
    final docRef = await _firestore.collection('movies').add(movie.toFirestore());
    // final generatedId = docRef.id;

    // Обновляем документ, чтобы сохранить сгенерированный id в объекте
    // await docRef.update({'id': generatedId});
  }

}